### PR TITLE
feat: Allow returning objects in `Parse.Cloud.beforeFind` without invoking database query

### LINKED ISSUE #8693: Allow beforeFind to be short circuited
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation

There are cases where it would be good to be able to return objects from a `beforeFind` trigger, bypassing the inbuilt `RestQuery`

### Feature / Enhancement Description

Allow beforeFind to return objects

### Example Use Case

- loading objects from a cache
- loading objects from an external search, such as Elastic Search
- gracefully cancelling a query, returning `[]` without needing to throw

### Alternatives / Workarounds

Use a cloud function

### 3rd Party References

https://community.parseplatform.org/t/return-the-data-from-beforefind-trigger-and-stop-the-find-query/826

### PR DESCRIPTION
**Pull request by EmpiDev**
## Pull Request

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

## Issue

Closes: https://github.com/parse-community/parse-server/issues/8693

## Approach

This PR continues and supersedes #8694 by @dblythy.
Adds ability to return objects (or an empty array) from a beforeFind trigger.

## Tasks

- [x] Add tests

## Summary by CodeRabbit

- **New Features**
  - find/get now support trigger-returned objects, including returning unsaved objects and get-specific behavior.

- **Bug Fixes**
  - Access control enforced when triggers return objects; protected fields remain masked after re-filtering.
  - Role/security checks run before trigger processing to avoid unexpected database access.

- **Tests**
  - Added extensive tests for beforeFind/afterFind flows, ACL/security, protectedFields, get-vs-find, and various query/input edge cases.

- **Refactor**
  - Centralized trigger and query handling; object class names preserved when converting results.

### CODE REVIEW

**Comment by parseplatformorg** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/acinader/pr-checks/60f0559b-28d7-4b8e-b114-cec45de4b4ac?refs=852f29f96fb458ac20c6ee31343a9e5f4514d847&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by mtrezza** [comment]
Started CI...

**Comment by EmpiDev** [comment]
Hi @mtrezza, just wondering if there's anything else I should do to trigger a re-run of the tests?

**Comment by mtrezza** [comment]
@EmpiDev You don't have permission to trigger a CI run. Once your first PR has been merged, the CI will run automatically.

**Comment by EmpiDev** [comment]
Hi @mtrezza,

I'm still encountering some test failures when running `npm test`.
On the `feat/before-found` branch, **3 tests are failing**:

```
BRANCH: FEAT BEFORE FOUND

3300 specs, 3 failures, 73 pending specs
Finished in 382.402 seconds
Randomized with seed 83016

Failures:

1) httpRequest should fail gracefully
   - Timeout - Async function did not complete within 10000ms

2) Server Url Checks does not have unhandled promise rejection in the case of load error
   - Timeout - Async function did not complete within 10000ms

3) Non-spec failure
   - Spec "httpRequest should fail gracefully" ran a "toBeRejected" expectation after it finished.
     → Did you forget to `return` or `await` the result of `expectAsync`?
     → Was `done()` invoked before an async operation completed?
     → Did an expectation follow a call to `done()`?
```

I noticed that these same tests are also failing on the `alpha` branch.
Is that expected behavior, or should they pass there?

Thanks in advance for your help!
Let me know if you want me to dig further.

**Comment by mtrezza** [comment]
These may be flaky tests. I'd have to re-run the CI to see whether they pass. If a test passes locally on your machine but fails in the CI with a timeout, then it's likely flaky. I've restarted the CI.

File: spec/CloudCode.spec.js [hunk 1]
```diff
@@ -1496,9 +1552,9 @@ describe('Cloud Code', () => {
   });
 
   /*
-    TODO: fix for Postgres
-    trying to delete a field that doesn't exists doesn't play nice
-   */
+      TODO: fix for Postgres
+      trying to delete a field that doesn't exists doesn't play nice
+     */
```
Comment (mtrezza) [inline spec/CloudCode.spec.js]: Please undo these changes.

File: spec/CloudCode.spec.js [hunk 2]
```diff
@@ -2004,7 +2060,6 @@ describe('Cloud Code', () => {
     }
   });
 });
-
```
Comment (mtrezza) [inline spec/CloudCode.spec.js:2007]: Undo

File: spec/CloudCode.spec.js [hunk 3]
```diff
@@ -3347,20 +3411,23 @@ describe('beforeLogin hook', () => {
...
-      expect(req.context).toBeDefined();
-    });
+  it_id('5656d6d7-65ef-43d1-8ca6-6942ae3614d5')(it)(
+    'should have expected data in request in beforeLogin',
+    async done => {
+      Parse.Cloud.beforeLogin(req => {
+        expect(req.object).toBeDefined();
+        expect(req.user).toBeUndefined();
+        expect(req.headers).toBeDefined();
+        expect(req.ip).toBeDefined();
```
Comment (mtrezza) [inline spec/CloudCode.spec.js]: Are these only style changes? If so, please undo. There seem to be lots of style changes, of unrelated code. Please undo that all so we can review the code that is relevant in this PR.

File: src/rest.js [hunk 4]
```diff
@@ -23,36 +23,91 @@ function checkTriggers(className, config, types) {
...
     context,
+    runBeforeFind: false,
   });
+
   return query.execute();
+}
+
+// Returns a promise for an object with optional keys 'results' and 'count'.
+const find = async (config, auth, className, restWhere, restOptions, clientSDK, context) => {
+  enforceRoleSecurity('find', className, auth);
```
Comment (Moumouls) [inline src/rest.js:126]: question: how the enforce role security was implemented before, it's not clear

File: src/rest.js [hunk 5]
```diff
@@ -23,36 +23,91 @@ function checkTriggers(className, config, types) {
...
+async function runFindTriggers(
+  config,
+  auth,
+  className,
+  restWhere,
+  restOptions,
+  clientSDK,
+  context,
+  isGet
+) {
```
Comment (Moumouls) [inline src/rest.js:35]: polish: i can suggest to switch to object interface, because at usage (ex: the isGet) with land with a not truly understanble "true" or "flase".

Having something "isGet: true" is easier to review and maintain, could potentially avoid bugs

File: src/rest.js [hunk 6]
```diff
@@ -23,36 +23,91 @@ function checkTriggers(className, config, types) {
...
+
+    const afterFindProcessedObjects = await triggers.maybeRunAfterFindTrigger(
+      triggers.Types.afterFind,
+      auth,
+      className,
+      objectsFromBeforeFind,
+      config,
+      new Parse.Query(className).withJSON({ where: restWhere, ...restOptions }),
+      context
+    );
```
Comment (Moumouls) [inline src/rest.js:101]: question: the after find trigger doesn't have the knowledge of "isGet" ?

File: src/rest.js [hunk 7]
```diff
@@ -23,36 +23,91 @@ function checkTriggers(className, config, types) {
...
+  restWhere = result.restWhere || restWhere;
+  restOptions = result.restOptions || restOptions;
+
+  if (result?.objects) {
+    const objectsFromBeforeFind = result.objects;
+
+    const afterFindProcessedObjects = await triggers.maybeRunAfterFindTrigger(
+      triggers.Types.afterFind,
+      auth,
+      className,
```
Comment (Moumouls) [inline src/rest.js:95]: issue: i suggest to introduce a new context variable like "beforeFindReturnObjects" to allow developers to adjust afterFind or skip afterFind logic if beforeFind returned objects

File: src/rest.js (hunk 5 above)
Comment (EmpiDev) [reply src/rest.js:35]: Yes, you're right. I've made the fix

File: src/rest.js (hunk 6 above)
Comment (EmpiDev) [reply src/rest.js:101]: It now has access to isGet.

File: spec/CloudCode.spec.js [hunk 8]
```diff
@@ -201,6 +201,226 @@ describe('Cloud Code', () => {
...
   });
+  it('beforeFind can return object without DB operation', async () => {
+    Parse.Cloud.beforeFind('beforeFind', () => {
+      return new Parse.Object('TestObject', { foo: 'bar' });
+    });
+    Parse.Cloud.afterFind('beforeFind', req => {
+      expect(req.objects).toBeDefined();
+      expect(req.objects[0].get('foo')).toBe('bar');
+    });
+    const newObj = await new Parse.Query('beforeFind').first();
```
Comment (mtrezza) [inline spec/CloudCode.spec.js]: Please rename the class like they are usually named in tests, I think `TestObject` or so. Naming an object class "beforeFind" may be confusing, especially for AI agents.

File: spec/CloudCode.spec.js [hunk 9]
```diff
@@ -201,6 +201,226 @@ describe('Cloud Code', () => {
       done();
     }
   });
+  it('beforeFind can return object without DB operation', async () => {
```
Comment (mtrezza) [inline spec/CloudCode.spec.js]: This tests "without DB op", but where is that verified?

Comment (EmpiDev) [reply spec/CloudCode.spec.js:204]: I've made the change

File: spec/CloudCode.spec.js (hunk 8 above)
Comment (EmpiDev) [reply spec/CloudCode.spec.js]: I've made the change

File: spec/CloudCode.spec.js (hunk 9 above)
Comment (EmpiDev) [reply spec/CloudCode.spec.js]: Yes, you're right — I updated the tests to include the verification.
We use a spy on databaseAdapter.find to ensure that no database query is executed.

File: spec/CloudCode.spec.js [hunk 10]
```diff
@@ -201,6 +201,267 @@ describe('Cloud Code', () => {
...
+    const config = Config.get('test');
```
Comment (mtrezza) [inline spec/CloudCode.spec.js]: There is a lot of duplicate code in these tests. Can you please extract the repetitive parts?

Comment (EmpiDev) [reply spec/CloudCode.spec.js]: I added a helper function for the beforeFind tests to reduce repetitive code

Comment (mtrezza) [reply spec/CloudCode.spec.js]: Repetitive code should go into the `beforeEach` of the `describe` group (without function). If it's not needed for all tests and all the tests are for a distinct feature or functionality, then create a new `describe` sub-group, otherwise the helper function should go to the top within the `describe˚ group, so it's easily visible for developers that it exists.

Comment (EmpiDev) [reply spec/CloudCode.spec.js]: Okay, I've made the change

**Comment by mtrezza** [comment]
@Moumouls would you kindly review the changes made after your feedback and close the feedback that you see as addressed?

File: spec/CloudCode.spec.js [hunk 11]
```diff
@@ -202,6 +202,248 @@ describe('Cloud Code', () => {
...
+      Parse.Cloud.afterFind('TestObject', req => {
+        expect(req.objects).toBeDefined();
+        expect(req.objects[0].get('foo')).toBe('bar');
+      });
+
+      const newObj = await new Parse.Query('TestObject').first();
+      expect(newObj.className).toBe('TestObject');
+      expect(newObj.toJSON()).toEqual({ foo: 'bar' });
+      expect(findSpy).not.toHaveBeenCalled();
+      await newObj.save();
```
Comment (Moumouls) [inline spec/CloudCode.spec.js:227]: suggestion: i think we need some additional use cases about security on objects returned in a before find, what is the behavior ?

If the code return object with ACL to user B for user A does parse return the full object, does parse filter again on ACL ?

to be honest i'm not 100% confident on the implementation code, because of some usage of instance of, feature detection, mutation of objects, and fragmented changes, if we can add some comments in the code it could helps for future maintenance/debug

**Review by Moumouls — COMMENTED** [review]
@mtrezza ping

File: spec/CloudCode.spec.js (hunk 11 above)
Comment (EmpiDev) [reply spec/CloudCode.spec.js:227]: **Answer to your questions:**
- If `beforeFind` returns a `Query` → the normal flow (`RestQuery`) continues with ACL/CLP applied.
- If `beforeFind` returns objects (`Parse.Object` or an array) → the DB is short-circuited: `afterFind` is executed, **but ACL/CLP/protectedFields are not re-applied**.

**If I understand your message correctly, the actions I should take are:**

1. Add explicit comments in `runFindTriggers`, `RestQuery`, and `triggers` to document the flow and potential risks.
2. A proposed security fix for the “short-circuit” case:
    - Re-apply ACL/CLP systematically when `beforeFind` returns objects (except if `useMasterKey` is used).

3. Add security test cases:
    - `beforeFind` returns objects not accessible to the current user → they must not be visible.
    - `protectedFields` are correctly hidden after re-filtering.

Comment (Moumouls) [reply spec/CloudCode.spec.js:227]: Hi @EmpiDev , yes it could be a step forward, i think enforcing ACL protection even when DB is short circuited is important to prevent potential security risks and unwanted data leak.

File: spec/CloudCode.spec.js [hunk 12]
```diff
@@ -202,6 +202,313 @@ describe('Cloud Code', () => {
...
+        className,
+        [mockObject],
+        testConfig,
+        { limit: 5, skip: 0 },
+        {}
+      );
+
+      expect(receivedQuery).toBeInstanceOf(Parse.Query);
+      expect(result).toBeDefined();
+    });
```
[the remark this answers is not in the document]
Comment (EmpiDev) [reply spec/CloudCode.spec.js:502]: 1. **The current behavior of `withJSON` already supports top-level options**

   * `withJSON` expects a full Parse Query JSON object (`{ where, limit, skip, order, etc. }`).
   * If the incoming object is already in that format, `withJSON` handles it automatically.
   * If the object is just a filter (no `where`), the current code wraps it in `{ where: query }`, which is exactly what you want.

2. **The proposed “TOP\_LEVEL” extraction is brittle**

   * Hardcoding keys like `limit`, `skip`, `order`, etc. introduces maintenance overhead.
   * If Parse adds a new top-level option, the list must be updated, creating unnecessary technical debt.

File: src/triggers.js [hunk 13]
```diff
@@ -437,69 +445,93 @@ function logTriggerErrorBeforeHook(triggerType, className, input, auth, error, l
...
-  className,
-  objects,
+  classNameQuery,
+  objectsInput,
   config,
   query,
-  context
+  context,
+  isGet
 ) {
```
[the remark this answers is not in the document]
Comment (EmpiDev) [reply src/triggers.js:454]: If all my tests had already passed before your change, that means it was okay, right

File: src/triggers.js [hunk 14]
```diff
@@ -437,69 +445,93 @@ function logTriggerErrorBeforeHook(triggerType, className, input, auth, error, l
...
       request.query = query;
+    } else if (typeof query === 'object' && query !== null) {
+      const parseQueryInstance = new Parse.Query(classNameQuery);
+      if (query.where) {
+        parseQueryInstance.withJSON(query);
+      }
+      request.query = parseQueryInstance;
+    } else {
+      request.query = new Parse.Query(classNameQuery);
     }
```
[the remark this answers is not in the document]
Comment (EmpiDev) [reply src/triggers.js:477]: This isn’t a real fix — the issue only appears in tests where we pass raw options; in production maybeRunAfterFindTrigger always gets a Parse.Query (via withJSON), so the bug doesn’t occur, no ?

File: src/rest.js [hunk 15]
```diff
@@ -23,36 +23,135 @@ function checkTriggers(className, config, types) {
...
+      objectsForAfterFind,
+      config,
+      new Parse.Query(className).withJSON({ where: restWhere, ...restOptions }),
+      context,
+      isGet
+    );
+
+    return {
+      results: afterFindProcessedObjects,
+    };
```
[the remark this answers is not in the document]
Comment (EmpiDev) [reply src/rest.js:105]: But what we want with this PR is to allow beforeFind to short-circuit by returning objects directly, while ensuring ACL/CLP security is applied no ?

File: spec/CloudCode.spec.js (hunk 11 above)
Comment (EmpiDev) [reply spec/CloudCode.spec.js:227]: Hi @Moumouls, I think I’ve made all the necessary changes. Could you check ?

File: src/rest.js [hunk 16]
```diff
@@ -23,36 +23,129 @@ function checkTriggers(className, config, types) {
...
+
+    let objectsForAfterFind = objectsFromBeforeFind;
+
+    // Security check: Re-filter objects if not master to ensure ACL/CLP compliance
+    if (!auth?.isMaster && !auth?.isMaintenance) {
+      const ids = (Array.isArray(objectsFromBeforeFind) ? objectsFromBeforeFind : [objectsFromBeforeFind])
+        .map(o => (o && (o.id || o.objectId)) || null)
+        .filter(Boolean);
+
+      if (ids.length > 0) {
```
Comment (Moumouls) [inline src/rest.js:69]: Issue: I think there’s a security concern here. In the case where no IDs are found (for example, if a developer returns objects without an .id — don’t ask me why 😅), the code skips the filter and does not reassign objectsForAfterFind.

Is this expected? Should we instead assign an empty array when ids.length === 0?

Pros:
- Allows the developer to return a non-existing object

Cons:
- Bypasses security

Comment (Moumouls) [reply src/rest.js:69]: @EmpiDev ping, and nice job for the ACL/CLP validation, elegant approach !

Comment (IlyaDiallo) [reply src/rest.js:69]: Objects without IDs are (normally) unsaved objects, and I think that the specs should allow us to return them, as beforeSave already does (see previous discussion in the original issue).
If the object is unsaved, I guess that the ACL security does not apply.

Comment (Moumouls) [reply src/rest.js:69]: Okay a comment should be added in the code 🚀

Comment (Moumouls) [reply src/rest.js:69]: @EmpiDev if you have some time to just push the comment and i believe we are good to go here, thanks for your contribution !

Comment (EmpiDev) [reply src/rest.js:69]: I've added the comments thanks to your help @Moumouls, really appreciate it

**Review by Moumouls — APPROVED** [review]
LGTM

File: src/rest.js (hunk 16 above)
Comment (Moumouls) [reply src/rest.js:69]: LGTM @EmpiDev nice job !

**Comment by mtrezza** [comment]
Started CI

**Comment by Moumouls** [comment]
everything seems okay here @mtrezza  :)

**Comment by mtrezza** [comment]
Merged alpha branch to run CI again;

@EmpiDev could you please resolve the lint issues?
@Moumouls could you please resolve any open comments - there are some hidden in the thread.

**Comment by Moumouls** [comment]
@mtrezza i don't see some unresolved comments, or i can't resolve them because file are outdated or i approved the PR already

**Comment by EmpiDev** [comment]
@mtrezza I have resolved the lint issues.

File: spec/CloudCode.spec.js (hunk 11 above)
Comment (mtrezza) [reply spec/CloudCode.spec.js:227]: @Moumouls Can this be resolved?

File: src/rest.js (hunk 7 above)
Comment (mtrezza) [reply src/rest.js:95]: @Moumouls Can this be resolved?

File: src/rest.js (hunk 4 above)
Comment (mtrezza) [reply src/rest.js:126]: @Moumouls Can this be resolved?

**Review by mtrezza — COMMENTED** [review]
@Moumouls I have found these open comments:

https://github.com/parse-community/parse-server/pull/9770#discussion_r2376234340

https://github.com/parse-community/parse-server/pull/9770#discussion_r2376241854

https://github.com/parse-community/parse-server/pull/9770#discussion_r2376244158

If you cannot resolve them because the button doesn't work, could you just confirm that they're resolved, so we can merge?

File: spec/CloudCode.spec.js (hunk 11 above)
Comment (Moumouls) [reply spec/CloudCode.spec.js:227]: yes it's resolved @mtrezza !

File: src/rest.js (hunk 7 above)
Comment (Moumouls) [reply src/rest.js:95]: maybe out of scope, so let's resolve this @mtrezza my comment "issue:" is wrong should be "suggestion:"

File: src/rest.js (hunk 4 above)
Comment (Moumouls) [reply src/rest.js:126]: Okay i investigated the code deeper it seems that in the common cases "enforceRoleSecurity" will run twice (here and in RestQuery instance), enforceRoleSecurity ONLY prevent some queries on core classes, it's a small sync function with 2 "if" so i guess it's okay to run it twice. What do you think @mtrezza  ?

@EmpiDev  you added this function here because some existing tests fails without it ?

**Comment by Moumouls** [comment]
thanks @mtrezza , i reviewed, there is still one open comment: https://github.com/parse-community/parse-server/pull/9770#discussion_r2388521470

File: src/rest.js (hunk 4 above)
Comment (EmpiDev) [reply src/rest.js:126]: It was added by @dblythy .
 As you correctly identified, this is indeed a duplication - enforceRoleSecurity runs twice in the common case: once here in find()/get(), and again in the RestQuery constructor.

I've verified that all tests pass without these calls in rest.js (lines 126 and 141), since the check is already performed in RestQuery() which is always called downstream.

I think we have two reasonable options:
- Keep it as an extra safety layer 
- Remove it to eliminate the duplication since RestQuery() already handles it

I'm fine with either approach. What's your preference @mtrezza?

Comment (Moumouls) [reply src/rest.js:126]: Since this security layer is super lightweight no performance impact will be measurable, i think we are done here

**Comment by mtrezza** [comment]
Waiting for CI...

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [8.3.0-alpha.5](https://github.com/parse-community/parse-server/releases/tag/8.3.0-alpha.5)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [8.3.0](https://github.com/parse-community/parse-server/releases/tag/8.3.0)
